### DISCUSSION TITLE
Managing Workflow between agent

### DISCUSSION THREAD

**Discussion by ggae2** [microsoft/autogen#4865] [General]
Hello 

I've researched and reviewed various articles about managing workflows between agents in AutoGen, but I'm still seeking clarity. Let me describe a specific use case:
We want to implement an AI Client Relationship Manager ( RM) that serves as the primary point of contact with clients. This manager would:

Interface directly with clients to understand their needs
Route conversations to specialized teams based on the issue type

For example:

If a client reports travel booking issues, the RM will ask genereic information and forwarded it to Team A (Travel Booking Specialists)
If a client mentions hotel booking problems, the RM will send the request Team B (Hotel Booking Specialists)

- in some case, the Teams A or B may request specific additional information from clients through the Relationship Manager
- The Relationship Manager to maintain its role as the primary client interface
- Specialized teams to work in the background without direct client interaction

I've attempted using group chat functionality, but found it inefficient as it involves all agents simultaneously. Instead, I wonder if ther's a solution that engages only the relevant team and agents at the appropriate time.

**Comment by ekzhu** [comment]
It's not something easy to do in v0.2 API. The v0.4 API is suitable for this, and it will be released officially as soon as next week. 

Please check out the followings in the v0.4 documentation:

1. Swarm (AgentChat): https://microsoft.github.io/autogen/0.4.0.dev13/user-guide/agentchat-user-guide/tutorial/swarm.html. This is still sharing all messages across all agents. But in the future with the addition of the memory module, we can start to perform selective filtering of messages within individual agents to achieve isolation. #4438 

2. Handoff Workflows (Core): https://microsoft.github.io/autogen/0.4.0.dev13/user-guide/core-user-guide/design-patterns/handoffs.html. This is using a lower level API but offers much more flexibility in how you would like to implement your workflow. So, you can create your customized workflow using the documented workflow as blueprint. Some ramp-up is needed to understand topic and subscription.

**Reply by cbsingh05iitkgp** [reply]
I am waiting for the release where it is covered. I have similar usecase , but not easy to implement currently

**Reply by cbsingh05iitkgp** [reply]
Please share example code as well. I am new to Autogen and Gen AI functionality

**Reply by ggae2** [reply]
Actually, I tried swarmagent and it cover a part of my need but I think a graph would be also the best option

**Reply by ekzhu** [reply]
There are two concepts in my original post:

(1) A group chat. This is the AgentChat API in which all agents share the same context. The underlying mechanism is based on message broadcast.

(2) A workflow. A more lower-level concept. It allows you create any kind of graph-like flows, using pub-sub. 

For (1), you use AgentChat. For (2) you use Core API. You can visit the documentation and start reading the basics from each. 

> Please share example code as well. I am new to Autogen and Gen AI functionality

For (2), there isn't any gen-AI, it is just event-driven programming. 

> Actually, I tried swarmagent and it cover a part of my need but I think a graph would be also the best option

if a graph is better, I would recommend start with the Core API to create a workflow based on your scenario. It is going to be lower level. High-level flow specification will come in the future.
